File: javascript/beer-song/beer-song.js
var BeerSong = function () {};

BeerSong.prototype.verse = function (num) {
  var line1 = num + ' bottles of beer on the wall, ';
  var line2 = num + ' bottles of beer.\n';
  var line3 = 'Take one down and pass it around, ';
  var line4 = (num - 1) + ' bottles of beer on the wall.\n';

  if (num === 0) {
    line1 = line1.replace('0', 'No more');
    line2 = line2.replace('0', 'no more');
    line3 = 'Go to the store and buy some more, ';
    line4 = line4.replace('-1', '99');
  } else if (num === 1) {
    line1 = line1.replace('bottles', 'bottle');
    line2 = line2.replace('bottles', 'bottle');
    line3 = line3.replace('one', 'it');
    line4 = line4.replace('0', 'no more');
  } else if (num === 2) {
    line4 = line4.replace('bottles', 'bottle');
  }

  return line1 + line2 + line3 + line4;
};

BeerSong.prototype.sing = function (from, to) {
  to = to || 0;
  
  var song = "";
  for (var i = from; i > to; i--) {
    song += this.verse(i) + '\n';
  }
  return song += this.verse(to);
};

module.exports = BeerSong;

File: ruby/beer-song/beer_song.rb
class BeerSong

  def verse(num)
    if num == 1
      "1 bottle of beer on the wall, 1 bottle of beer.\n" \
      "Take it down and pass it around, no more bottles of beer on the wall.\n"
    elsif num == 0
      "No more bottles of beer on the wall, no more bottles of beer.\n" \
      "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
    else
      "#{num} #{units(num)} of beer on the wall, #{num} #{units(num)} of beer.\n" \
      "Take one down and pass it around, #{num-1} #{units(num-1)} of beer on the wall.\n"
    end
  end

  def verses(first, last)
    verse_num = first
    verses = ""
    until verse_num < last
      verses += verse(verse_num)
      verses += "\n" unless verse_num == last
      verse_num -= 1
    end
    verses
  end

  def lyrics
     verses(99,0)
  end

  def units(num)
    num == 1 ? "bottle" : "bottles"
  end
end

module BookKeeping
  VERSION = 2
end
